src/intrusive_pairing_heap.rs

```diff
@@ -0,0 +1,402 @@
+    first_child: Option<NonNull<HeapNode<T>>>,
+    /// The data which is associated to this heap item.
+    data: T,
+    /// Prevents `HeapNode`s from being `Unpin`. They may never be moved, since
+    /// the heap semantics require addresses to be stable.
+    _pin: PhantomPinned,
+}
+
+impl<T> HeapNode<T> {
+    /// Creates a new node with the associated data
+    pub fn new(data: T) -> HeapNode<T> {
+        HeapNode::<T> {
+            parent: None,
+            prev: None,
+            next: None,
+            first_child: None,
+            data,
+            _pin: PhantomPinned,
+        }
+    }
+
+    fn is_root(&self) -> bool {
+        if self.parent.is_none() {
+            debug_assert_eq!(self.prev, None);
+            debug_assert_eq!(self.next, None);
+            true
+        } else {
+            false
+        }
+    }
+}
+
+impl<T> Deref for HeapNode<T> {
+    type Target = T;
+
+    fn deref(&self) -> &T {
+        &self.data
+    }
+}
+
+impl<T> DerefMut for HeapNode<T> {
+    fn deref_mut(&mut self) -> &mut T {
+        &mut self.data
+    }
+}
+
+/// Add a child to a node.
+unsafe fn add_child<T: Ord>(
+    mut parent: NonNull<HeapNode<T>>,
+    mut child: NonNull<HeapNode<T>>,
+) {
+    // require parent <= child
+    debug_assert!(!safe_lesser(&child.as_ref().data, &parent.as_ref().data));
+    if let Some(mut old_first_child) = parent.as_mut().first_child.take() {
+        child.as_mut().next = Some(old_first_child);
+        old_first_child.as_mut().prev = Some(child);
+    }
+    parent.as_mut().first_child = Some(child);
+    child.as_mut().parent = Some(parent);
+}
+
+/// Merge two root heaps. Returns the new root.
+unsafe fn meld<T: Ord>(
+    mut left: NonNull<HeapNode<T>>,
+    mut right: NonNull<HeapNode<T>>,
+) -> NonNull<HeapNode<T>> {
+    debug_assert!(left.as_ref().is_root());
+    debug_assert!(right.as_ref().is_root());
+    if safe_lesser(&right.as_ref().data, &left.as_ref().data) {
+        mem::swap(&mut left, &mut right);
+    }
+    // Now we always have `left` <= `right`. `left` will be the new root.
+    add_child(left, right);
+    left
+}
+
+/// Merge two root heaps, where the left might be empty. Returns the new root.
+unsafe fn maybe_meld<T: Ord>(
+    left: Option<NonNull<HeapNode<T>>>,
+    right: NonNull<HeapNode<T>>,
+) -> NonNull<HeapNode<T>> {
+    if let Some(left) = left {
+        meld(left, right)
+    } else {
+        right
+    }
+}
+
+/// Given the first child in a child list, traverse and find the last child.
+unsafe fn last_child<T>(
+    mut first_child: NonNull<HeapNode<T>>,
+) -> NonNull<HeapNode<T>> {
+    while let Some(next) = first_child.as_ref().next {
+        first_child = next;
+    }
+    first_child
+}
+
+unsafe fn unlink_prev<T>(
+    mut node: NonNull<HeapNode<T>>,
+) -> Option<NonNull<HeapNode<T>>> {
+    let mut prev = node.as_mut().prev.take()?;
+    debug_assert_eq!(prev.as_ref().next, Some(node));
+    prev.as_mut().next = None;
```

Comment:
What is with the previous nodes before the unlinked one? Do we need to adjust pointers?

```
prev.prev.next = Some(next);
```` 

and what about parent/child pointers on the unlinked node?

Reply:
`prev.prev` shouldn't be touched; the idea is that given a list of the form
```
A <-> B <-> C
```
calling `unlink_prev` on `C` should mutate the list to become
```
A <-> B      C
```
and return `B`.

Parent/child pointers aren't touched by this function.

I'll add a comment.